### ISSUE TITLE
Improvements to  Library View

### ISSUE THREAD

**Issue by JamesSwift** [johnfactotum/foliate#396]
Hi @johnfactotum. First off, congreats on the latest release. It's a huge improvement, and much appreciated! Thank you for all your hard work.

I was wondering if it would be possible to alter the library view in the following ways:

 - Allow sorting by name as well as most recent
 - Have an option to always show all books (instead of having to press the show more button)
 - Have an option to show the title underneath the book (as some cover's aren't very readable)
 - Allow some kind of grouping (possibly by which folder the file is in), to make it easier to find files in a large library

Also, I'd like to donate towards development costs, but I'm not sure how...

**Comment by johnfactotum** [comment]
> Allow sorting by name as well as most recent

The main technical reason this isn't implemented is that this requires reading all the contents of all the JSON files, while sorting by most recent requires only listing the files and querying the modified time. While it's true that when searching, it will read all the files anyway, when not searching this improves performance.

Design-wise, the use for sorting by name is slightly limited because you can just search. This might not be a very compelling reason but it is something to consider.

> Have an option to always show all books (instead of having to press the show more button)

Again this is mainly for performance reasons. But I suppose infinite scrolling would be easier to use than having to click on load more.

> Have an option to show the title underneath the book (as some cover's aren't very readable)

Agreed, but maybe it shouldn't be optional. We don't want too many options.

> Allow some kind of grouping (possibly by which folder the file is in), to make it easier to find files in a large library

We can probably add collections. Personally I don't think it makes a lot of sense to group by the folder of the file, since the whole point of a "library" is so that you don't have to think about the file system -- for that one can just use any file manager. Plus in Flatpak we don't know the real location of the file.

**Comment by digitalethics** [comment]
> Allow some kind of grouping (possibly by which folder the file is in)

> We can probably add collections. Personally I don't think it makes a lot of sense to group by the folder of the file

I think it'll be good to assign _labels_ to ebooks (e.g. philosophy, science, literature, etc.) and to allow arranging them by labels.

**Comment by johnfactotum** [comment]
Yeah, that's what I meant by "collections", which is what it's called in most e-book apps.

**Comment by JamesSwift** [comment]
> > Allow sorting by name as well as most recent
> 
> The main technical reason this isn't implemented is that this requires reading all the contents of all the JSON files, while sorting by most recent requires only listing the files and querying the modified time. While it's true that when searching, it will read all the files anyway, when not searching this improves performance.

You could have two options, sort by filename (which would be equally as quick), and sort by name. That would actually be even more handy for my particular use case. As to performance, there is unlikely to be much of a hit unless you have thousands of books, but in any case if you have "sort by recently used" as the default, then it's not a problem.


> Design-wise, the use for sorting by name is slightly limited because you can just search. This might not be a very compelling reason but it is something to consider.

I take your point, but still I feel it is useful. File managers have "sort by" for a reason :-)


> > Have an option to show the title underneath the book (as some cover's aren't very readable)
> 
> Agreed, but maybe it shouldn't be optional. We don't want too many options.

So default then? Options aren't a bad thing though...


> > Allow some kind of grouping (possibly by which folder the file is in), to make it easier to find files in a large library
> 
> We can probably add collections. Personally I don't think it makes a lot of sense to group by the folder of the file, since the whole point of a "library" is so that you don't have to think about the file system -- for that one can just use any file manager. Plus in Flatpak we don't know the real location of the file.

Fair enough. @digitalethics 's idea about labels would work for my usecase as long as it allows some manual tagging. I just thought using the directory would be quicker to implement.

**Comment by gregory-opera** [comment]
> Yeah, that's what I meant by "collections", which is what it's called in most e-book apps.

That's what I suggested over [here](https://github.com/johnfactotum/foliate/issues/570)... It would be great to be able to "group" a collection of eBooks, digital comics or magazines together, rather than having them in one huge list (which is how it currently works).

For example, I could "group" my eBooks together by series (e.g. "Odd Thomas"), my comics by story arc (e.g. "In the Spider-Verse") and my magazines by publication (e.g. Vetaffairs).

Currently, the lack of this _particular_ functionality is _almost_ a deal-breaker...

**Comment by jwiede** [comment]
> "Have an option to always show all books"

This particular bit of Folate behavior is driving me INSANE.  I have hundreds of ebooks, and having to press "show more" over and over (and over and over and over...) to be able to see everything present is a HUGE pain point for me.  As much as I like Foliate otherwise, the current behavior of only showing books in "batches" (where size doesn't even seem controllable) is incredibly frustrating when I need books that have wound up towards the end of the set (regardless of ordering).

Please, PLEASE add an option where Foliate can scan/cache all volumes "added" -- if it takes a while, fine, have a progress graph (you should know how many books I've added total), but having to press "show more" a dozen times or more every time I go back to the library to see books in the middle to end of the displayed set is... INFURIATING, to be polite.  Deeply frustrating and happens ALL the time as a daily part of my Foliate UX.  Just saying.

Thanks as always, I LOVE Foliate, and it's still my main ebook reader on Linux, but the more ebooks I move here, the more Foliate itself seems to be driving me to find something with marginally less painful UX for library browsing.  I really want Foliate to succeed, and this seems likely to be viewed as a massive UX problem by ANY user that has loads of ebooks, unfortunately.

**Comment by johnfactotum** [comment]
@jwiede The issue is currently tracked in #583. One bottleneck currently is GtkFlowBox's scalability. Once we port to GTK 4, we should be able to display a virtually unlimited number of books at once (I think).

**Comment by Delta-official** [comment]
any updates on this?

**Comment by johnfactotum** [comment]
>  Allow sorting by name as well as most recent

No update on this.

> Have an option to always show all books (instead of having to press the show more button)

Fixed in the gtk4 branch. See #583.

> Have an option to show the title underneath the book (as some cover's aren't very readable)

Fixed in the gtk4 branch.

> Allow some kind of grouping (possibly by which folder the file is in), to make it easier to find files in a large library

No update on this. See #570.
